File: 03-Arrays-Part-III/python/02_pow_x_n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        power = n
        if power < 0:
            x = 1 / x
            power = -power

        result = 1.0
        current_product = x

        while power > 0:
            if power & 1:
                result *= current_product
            current_product *= current_product
            power >>= 1

        return result

File: 03-Arrays-Part-III/cpp/02-pow-x-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
        long long power = n;
        if (power < 0) {
            x = 1 / x;
            power = -power;
        }

        double result = 1.0;
        double currentProduct = x;

        while (power > 0) {
            if (power & 1) {
                result *= currentProduct;
            }
            currentProduct *= currentProduct;
            power >>= 1;
        }

        return result;
    }
};
